### PR TITLE
Fix flush

### PR DESCRIPTION
**Pull request by craff**

### CODE REVIEW

File: src/ssl_stubs.c [hunk 1]
```diff
@@ -1751,8 +1751,11 @@ CAMLprim value ocaml_ssl_flush(value socket)
   bio = SSL_get_wbio(ssl);
   if(bio)
   {
-    /* TODO: raise an error */
-    assert(BIO_flush(bio) == 1);
+    int ret = BIO_flush(bio);
+    if (ret != 1) {
```
Comment (anmonteiro) [inline src/ssl_stubs.c:1755]: I think this needs to acquire the runtime lock again before raising.

File: src/ssl.ml [hunk 2]
```diff
@@ -96,6 +96,7 @@ exception Accept_error of ssl_error
 exception Read_error of ssl_error
 exception Write_error of ssl_error
 exception Verify_error of verify_error
+exception Flush_error of bool (* true means retry *)
```
Comment (anmonteiro) [inline src/ssl.ml:99]: This should be added to the `.mli`

Comment (craff) [reply src/ssl.ml:99]: Thanks, just did it

File: src/ssl_stubs.c (hunk 1 above)
Comment (craff) [reply src/ssl_stubs.c:1755]: done too.

**Comment by craff** [comment]
By the way, the tread_safe parameter to Ssl.init should be enough ? Why to keep a separate Ssl_thread module with nothing in it ? I will do another PR.

**Comment by craff** [comment]
And thanks for your review.

**Comment by anmonteiro** [comment]
> Why to keep a separate Ssl_thread module with nothing in it ?

Probably for backwards compatibility. It's a breaking change to remove it, and will likely break some OPAM code. An analysis of the opam repository should probably be done first.

**Comment by craff** [comment]
We can keep Ssl_thread.init, mark it as deprecated and make Ssl.init ~thread_safe true position the reference
so that it is sufficient to be thread safe?

**Comment by craff** [comment]
I fixed the problem of compilation on 5.0 without using Thread.Exit. So now it compiles with both old OCaml and 5.0. test shoud work.

**Comment by craff** [comment]
My previous comment was unclear: ocaml-ssl did not compile on 5.0 because of Thread.exit () deprecated and warn as error. My first fix was bad (broke even 4.13!), I just added a "+w 3" in dune for tests/util.ml to keep the warning and not make it an error.

**Comment by craff** [comment]
The build on macos failed to download alcotest for a strange reason. Probably a temporary failure.

**Comment by craff** [comment]
replaced by #101

File: src/ssl.mli [hunk 3]
```diff
@@ -96,6 +96,10 @@ exception Read_error of ssl_error
 (** An error occured while writing data. *)
 exception Write_error of ssl_error
 
+(** An error occured while flushing a socket.
+    [Flush_error true],  means that the operation should be retried later. *)
```
Comment (anmonteiro) [inline src/ssl.mli:100]:
```suggestion
(** An error occured while flushing a socket.
    [Flush_error true] means that the operation should be retried. *)
```

File: tests/dune [hunk 4]
```diff
@@ -1,6 +1,7 @@
 (library
  (name util)
  (modules util)
+ (flags -w +3)
```
Comment (anmonteiro) [inline tests/dune:4]: - isn't warning 3 turned on by default already?
- this call is dropping other flags, it should read `(flags :standard -w .. )`

Comment (craff) [reply tests/dune]: Yes, you are right. A better fix of the compilation problem is and [@warning] that is in commit [4690761](https://github.com/savonet/ocaml-ssl/pull/104/commits/46907610c2b908368a9d8f8eb9e592f0f660e593)

File: src/ssl.mli [hunk 5]
```diff
@@ -96,6 +96,10 @@ exception Read_error of ssl_error
 (** An error occured while writing data. *)
 exception Write_error of ssl_error
 
+(** An error occured while flushing a socket.
```
Comment (anmonteiro) [inline src/ssl.mli:99]:
```suggestion
(** An error occurred while flushing a socket.
```

**Review by anmonteiro — APPROVED** [review]
Thanks

**Pull request merged**